eth/downloader/downloader.go

```diff
@@ -1864,3 +1873,70 @@ func (d *Downloader) deliver(id string, destCh chan dataPack, packet dataPack, i
 		return errNoSyncActive
 	}
 }
+
+// qosTuner is the quality of service tuning loop that occasionally gathers the
+// peer latency statistics and updates the estimated request round trip time.
+func (d *Downloader) qosTuner() {
+	for atomic.LoadInt32(&d.interrupt) == 0 {
+		// Retrieve the current median RTT and integrate into the previoust target RTT
+		rtt := time.Duration(float64(1-qosTuningImpact)*float64(atomic.LoadUint64(&d.rttEstimate)) + qosTuningImpact*float64(d.peers.medianRTT()))
+		atomic.StoreUint64(&d.rttEstimate, uint64(rtt))
+
+		// A new RTT cycle passed, increase our confidence in the estimated RTT
+		conf := atomic.LoadUint64(&d.rttConfidence)
+		conf = conf + (1000000-conf)/2
+		atomic.StoreUint64(&d.rttConfidence, conf)
+
+		// Log the new QoS values and sleep until the next RTT
+		glog.V(logger.Debug).Infof("Quality of service: rtt %v, conf %.3f, ttl %v", rtt, float64(conf)/1000000.0, d.requestTTL())
+		time.Sleep(rtt)
```

Comment:
The downloader doesn't have a quit channel, only the interrupt. Also the time we sleep changes based on the RTT value (e.g. if RTT it 1 sec, it's pointless to wait for 20 second to tune, but it RTT is 20 sec, it's meaningless to tune it every sec... given that tuning converges the current value to the measured values, it should only happen when the measured value has a realistic change to change). A timer could be used, but not sure it that would matter much.